wire-library/wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/internal/grpc.kt

```diff
@@ -115,25 +129,40 @@ internal fun <R : Any> SendChannel<R>.readFromResponseBodyCallback(
  * 1. read a message (non blocking, suspending code)
  * 2. write it to the stream (blocking)
  * 3. repeat. We also have to wait for all 2s to end before closing the writer
+ *
+ * If this fails reading a message from the channel, we need to call [MessageSink.cancel]. If it
+ * fails writing a message to the network, we shouldn't call that method.
+ *
+ * If it fails either reading or writing we need to call [MessageSink.close] (via [Closeable.use])
+ * and [ReceiveChannel.cancel].
  */
 internal suspend fun <S : Any> ReceiveChannel<S>.writeToRequestBody(
   requestBody: PipeDuplexRequestBody,
   minMessageToCompress: Long,
   requestAdapter: ProtoAdapter<S>,
   callForCancel: Call
 ) {
-  requestBody.messageSink(minMessageToCompress, requestAdapter, callForCancel)
-    .use { requestWriter ->
-      var success = false
+  val requestWriter = requestBody.messageSink(minMessageToCompress, requestAdapter, callForCancel)
+  try {
+    requestWriter.use {
+      var channelReadFailed = true
       try {
         consumeEach { message ->
+          channelReadFailed = false
           requestWriter.write(message)
+          channelReadFailed = true
         }
-        success = true
+        channelReadFailed = false
       } finally {
-        if (!success) requestWriter.cancel()
+        if (channelReadFailed) requestWriter.cancel()
```

Comment:
@swankjesse You mean `channelWriteFailed`, yeah?

Reply:
No, you meant `channelReadFailed` because we're reading from the receivechannel of course